src/librustc_mir/transform/check_unsafety.rs

```diff
@@ -435,14 +433,23 @@ impl<'a, 'tcx> UnsafetyChecker<'a, 'tcx> {
                                         "mutating layout constrained fields cannot statically be \
                                         checked for valid values",
                                     )
-                                } else {
+
+                                // Check `is_freeze` as late as possible to avoid cycle errors
+                                // with opaque types.
```

Comment:
Do we have a test which would cause cycle errors if not for this commit?

Reply:
I wouldn't have fixed this otherwise. I don't think that it's a passing test though. (I'll go write one).